text/0000-inline-asm.md

```diff
@@ -0,0 +1,892 @@
+problems at hand.
+However, particularly where precise timing, and full control over the number of generated
+instructions is required intrinsics fall short.
+
+Intrinsics are of course still useful and have their place for inserting specific instructions.
+E.g. making sure a loop uses vector instructions, rather than relying on auto-vectorization.
+
+However, inline assembly is specifically designed for cases where more control is required.
+Also providing an intrinsic for every (potentially obscure) instruction that is needed
+e.g. during early system boot in kernel code is unlikely to scale.
+
+## Make the `asm!` macro return outputs
+
+It has been suggested that the `asm!` macro could return its outputs like the LLVM statement does.
+The benefit is that it is clearer to see that variables are being modified.
+Particular in the case of initialization it becomes more obvious what is happening.
+On the other hand by necessity this splits the direction and constraint specification from
+the variable name, which makes this syntax overall harder to read.
+
+```rust
+fn mul(a: u32, b: u32) -> u64 {
+    let (lo, hi) = unsafe {
+        asm!("mul {}", in(reg) a, in("eax") b, lateout("eax"), lateout("edx"))
+    };
+
+    hi as u64 << 32 + lo as u64
+}
+```
+
+# Prior art
+[prior-art]: #prior-art
+
+## GCC inline assembly
+
+The proposed syntax is very similar to GCC's inline assembly in that it is based on string substitution while leaving actual interpretation of the final string to the assembler. However GCC uses poorly documented single-letter constraint codes and template modifiers. Clang tries to emulate GCC's behavior, but there are still several cases where its behavior differs from GCC's.
+
+The main reason why this is so complicated is that GCC's inline assembly basically exports the raw internals of GCC's register allocator. This has resulted in many internal constraint codes and modifiers being widely used, despite them being completely undocumented.
+
+## D & MSVC inline assembly
+
+See the section [above][dsl].
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Flag outputs
+
+GCC supports a special type of output which allows an asm block to return a `bool` encoded in the condition flags register. This allows the compiler to branch directly on the condition flag instead of materializing the condition as a `bool`.
+
+We can support this in the future with a special output operand type.
+
+## `asm goto`
+
+GCC supports passing C labels (the ones used with `goto`) to an inline asm block, with an indication that the asm code may jump directly to one of these labels instead of leaving the asm block normally.
+
+This could be supported by allowing code blocks to be specified as operand types. The following code will print `a` if the input value is `42`, or print `b` otherwise.
```

Comment:
Alternatively, we could allow labels as arguments that can be jumped to, similar to https://github.com/rust-lang/rfcs/pull/2046?

Reply:
@RalfJung That would make it substantially less structured. I'm hoping we can handle all use cases via what looks like a single multi-way branch similar to an if.